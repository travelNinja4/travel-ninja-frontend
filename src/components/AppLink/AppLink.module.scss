/**
------------------------------------------------------
styles specific to the AppLink component
------------------------------------------------------
**/

.container {
  padding: 1rem;
  // Add styles here
}
