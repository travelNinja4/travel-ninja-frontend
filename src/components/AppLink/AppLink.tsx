/**
 * AppLink is a reusable wrapper around Next.jsâ€™s Link that standardizes navigation across the application.
 *
 * @example
 * ```tsx
 * import AppLink from '@src/components/AppLink'
 *
 * export default function AppLink() {
 *   return <AppLink href="/about">About</AppLink>;
 * }
 * ```
 */
'use client';

import { ReactNode } from 'react';
import Link from 'next/link';
import styles from './AppLink.module.scss';

/**
 * Define the props available for the AppLink component.
 */
interface AppLinkProps {
  /**
   * The destination URL of the link.
   *
   * For internal links, provide a relative path (e.g., `/about`).
   * For external links, provide a full URL (e.g., `https://example.com`).
   */
  href: string;

  /**
   * The content to be rendered inside the link.
   * Typically a string or React element such as text, icons, or both.
   */
  children: ReactNode;

  /**
   * When true, replaces the current history entry instead of pushing a new one.
   * Useful when you want to avoid keeping the previous page in the browser history.
   *
   * @default false
   */
  replace?: boolean;

  /**
   * Whether the link points to an external resource.
   *
   * If true, the link will open in a new tab (`target="_blank"`)
   * with security attributes (`rel="noopener noreferrer"`).
   */
  isExternal?: boolean;

  /**
   * Custom CSS class to apply to the rendered link element.
   * Useful for applying utility classes or module styles.
   */
  className?: string;
}

export default function AppLink({
  href,
  children,
  replace = false,
  isExternal = false,
  className,
  ...rest
}: AppLinkProps) {
  if (isExternal) {
    return (
      <a href={href} target="_blank" rel="noopener noreferrer" className={className} {...rest}>
        {children}
      </a>
    );
  }
  return (
    <Link href={href} replace={replace} className={className} data-testid="AppLinkTest">
      {children}
    </Link>
  );
}
