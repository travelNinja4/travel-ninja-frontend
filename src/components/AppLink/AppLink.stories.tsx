import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import AppLink from './AppLink';

const meta = {
  title: 'components/AppLink',
  component: AppLink,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    href: {
      control: 'text',
      description: 'Destination URL for the link',
    },
    children: {
      control: 'text',
      description: 'Content inside the link',
    },
    isExternal: {
      control: 'boolean',
      description: 'Opens the link in a new tab if true',
    },
    replace: {
      control: 'boolean',
      description: 'Replaces the current history state instead of pushing a new URL',
    },
    className: {
      control: 'text',
      description: 'Custom CSS class for styling',
    },
  },
  args: {
    href: '/about',
    children: 'About Us',
    isExternal: false,
  },
} satisfies Meta<typeof AppLink>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    href: '/about',
    children: 'Go to About',
  },
};

export const External: Story = {
  args: {
    href: 'https://google.com',
    children: 'Go to Google',
    isExternal: true,
  },
};

export const CustomClass: Story = {
  args: {
    href: '/contact',
    children: 'Contact Us',
    className: 'text-blue-600 underline hover:text-blue-800',
  },
};
