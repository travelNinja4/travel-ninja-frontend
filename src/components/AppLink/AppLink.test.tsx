import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import AppLink from './AppLink';

describe('Home', () => {
  /** Base props for AppLink **/
  const baseProps = {
    href: '/test',
    children: <span>AppLink Test</span>,
  };

  /**
   * @test
   * Renders the AppLink component without throwing an error
   * and checks if the AppLinkTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<AppLink {...baseProps} />);
    expect(screen.getByTestId('AppLinkTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the AppLink component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<AppLink {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the AppLink component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<AppLink {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders an external link with correct attributes', () => {
    render(
      <AppLink href="https://example.com" isExternal>
        External Link
      </AppLink>,
    );
    const link = screen.getByRole('link', { name: /external link/i });
    expect(link).toHaveAttribute('href', 'https://example.com');
    expect(link).toHaveAttribute('target', '_blank');
    expect(link).toHaveAttribute('rel', 'noopener noreferrer');
  });
});
