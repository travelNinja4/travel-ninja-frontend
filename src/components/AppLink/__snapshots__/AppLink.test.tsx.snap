// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <a
    data-testid="AppLinkTest"
    href="/test"
  >
    <span>
      AppLink Test
    </span>
  </a>
</DocumentFragment>
`;
